kernel/src/mm/physical_memory_manager.rs

```diff
@@ -113,80 +133,140 @@ impl PhysicalMemoryManager {
         }
     }
 
-    /// Look for `pages_needed` contiguous unused pages, beware of pages that are in use by the
-    /// kernel or reserved by opensbi.
-    fn find_contiguous_unused_pages(
+    fn exclude_range<const MAX_REGIONS: usize>(
+        regions: &mut [Option<AddressRange<usize>>; MAX_REGIONS],
+        excluded: (usize, usize),
+    ) {
+        let (excl_start, excl_end) = excluded;
+
+        assert!(excl_start < excl_end);
+
+        for i in 0..MAX_REGIONS {
+            if regions[i].is_none() {
+                continue;
+            }
+            let region = regions[i].unwrap();
+
+            if region.start == excl_start && region.end == excl_end {
+                // Perfect overlap between the region to be excluded and the current region, just remove the region.
+                regions[i] = None;
+            } else if (region.start < excl_start && excl_start < region.end)
+                && (region.start < excl_end && excl_end < region.end)
+            {
+                // Region to be excluded is the middle of the current region.
+                let new_region = AddressRange {
+                    start: excl_end,
+                    end: region.end,
+                };
+                regions[i] = Some(AddressRange::new(region.start, excl_start));
+
+                // The exclusion in the middle causes a split of the current region, put the new region (the end part) somewhere there is a none.
+                for j in 0..MAX_REGIONS {
```

Comment:
I agree with clippy here
```suggestion
                for region in regions.iter_mut() {
```
with `take(MAX_REGIONS)` added if that's what you need

Reply:
Done with this one, even you used nice functional stuff